### DISCUSSION TITLE
How to adjust output Quality and use other functions of ImageMagick with Magic-WASM?

### DISCUSSION THREAD

**Discussion by abduljabbar021** [dlemstra/magick-wasm#159] [Q&A] [answered]
First of all, it's an amazing tool. It works flawless with Workers.

I am looking to adjust/reduce the quality of output image (JPEG). I tried following but it throws error

        const MagickColor = entireMagickLibrary.MagickColor; 
        const response = await fetch(event.data.source);
        const buffer = await response.arrayBuffer();
        const data = new Uint8Array(buffer);
        entireMagickLibrary.ImageMagick.read(data, image => {
            image.quality(40); //Tried calling it here to reduce the image to 40
            image.write(entireMagickLibrary.MagickFormat.Png, data => {
                wk.postMessage(data);
            });
        });

But it throws error as 

       image.quality is not a function

I don't think this library is limited to only 4 functions as blur, resize, charcoal and rotate. If in case this library is limited to certain functions only, I couldn't find any list of all those functions other than 4 mentioned above.

Actually I want to lower/reduce the quality of images maintaining their real image size as we would do with ImageMagic
  
        convert input.jpg -quality 80 output.jpg

How can we use other functions like crop and other effects

**Comment by dlemstra** [comment, marked as answer]
The error message means that you are trying to execute something as a function that is not a function. This is happening because `quality` is a property and not a function. You will need to do `image.quality = 40;` instead.

This library does not have all functions implemented but it's a lot more than `4`. You can find which one are now implemented in this typescript file here: https://github.com/dlemstra/magick-wasm/blob/main/src/magick-image.ts

**Reply by abduljabbar021** [reply]
Thanks @dlemstra I didn't think it could be a property instead of a function. Stupid of me. Even though I went into implementation of magick-image.ts. Thank a lot for such awesome tool.

One more thing is that, is it possible for you to also include a command like function which takes exact ImageMagick command like
       
        convert input.jpg -quality 80 output.jpg

instead of implementing each method individually? Wouldn't that allow for more robust execution and direct access to ImageMagick?

**Reply by dlemstra** [reply]
That is by design. This exposes the internal API of ImageMagick and provides easier access to the available methods. And having to use convert would require some magic to read the files from disk because that is now not possible.

**Comment by Chew** [comment]
How did you get it to run on Workers? I tried everything but it will not function at all.

**Reply by young-energy** [reply]
same. how do you instantiate entireMagickLibrary ?
